src/app/reporting/LoggingManagement.cpp

```diff
@@ -0,0 +1,944 @@
+    {
+        eventDataElementBuilder.Number(apContext->mCurrentEventNumber);
+        eventDataElementBuilder.SystemTimestamp(apOptions->timestamp.systemTimestamp);
+    }
+    // else record delta
+    else
+    {
+        uint32_t deltatime = apOptions->timestamp.systemTimestamp - apContext->mCurrentTime;
+        eventDataElementBuilder.DeltaSystemTimestamp(deltatime);
+    }
+
+    SuccessOrExit(eventDataElementBuilder.GetError());
+
+    err = apContext->mWriter.StartContainer(ContextTag(chip::app::EventDataElement::kCsTag_Data), chip::TLV::kTLVType_Structure,
+                                            dataContainerType);
+    SuccessOrExit(err);
+    // Callback to write the EventData
+    err = apDelegate->WriteEvent(apContext->mWriter);
+    SuccessOrExit(err);
+
+    err = apContext->mWriter.EndContainer(dataContainerType);
+    SuccessOrExit(err);
+
+    eventDataElementBuilder.EndOfEventDataElement();
+    SuccessOrExit(eventDataElementBuilder.GetError());
+
+    err = apContext->mWriter.Finalize();
+    SuccessOrExit(err);
+
+    // only update mFirst if an event was successfully written.
+    if (apContext->mFirst)
+    {
+        apContext->mFirst = false;
+    }
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        apContext->mWriter = checkpoint;
+    }
+    else
+    {
+        // update these variables since BlitEvent can be used to track the
+        // state of a set of events over multiple calls.
+        apContext->mCurrentEventNumber++;
+        {
+            apContext->mCurrentTime = apOptions->timestamp.systemTimestamp;
+        }
+    }
+    return err;
+}
+
+void LoggingManagement::SkipEvent(EventLoadOutContext * apContext)
+{
+    apContext->mCurrentEventNumber++; // Advance the event id without writing anything
+}
+
+void LoggingManagement::CreateLoggingManagement(Messaging::ExchangeManager * apMgr, size_t aNumBuffers,
+                                                const LogStorageResources * const apLogStorageResources)
+{
+    new (&sInstance) LoggingManagement(apMgr, aNumBuffers, apLogStorageResources);
+}
+
+/**
+ * @brief Perform any actions we need to on shutdown.
+ */
+void LoggingManagement::DestroyLoggingManagement(void)
+{
+    Platform::CriticalSectionEnter();
+    sInstance.mState        = kLoggingManagementState_Shutdown;
+    sInstance.mpEventBuffer = NULL;
+    Platform::CriticalSectionExit();
+}
+
+CHIP_ERROR LoggingManagement::SetExchangeManager(Messaging::ExchangeManager * apMgr)
+{
+    mpExchangeMgr = apMgr;
+    return CHIP_NO_ERROR;
+}
+
+LoggingManagement::LoggingManagement(Messaging::ExchangeManager * apMgr, size_t aNumBuffers,
+                                     const LogStorageResources * const apLogStorageResources)
+{
+    CircularEventBuffer * current = NULL;
+    CircularEventBuffer * prev    = NULL;
+    CircularEventBuffer * next    = NULL;
+    size_t bufferIndex, j;
+
+    VerifyOrDie(aNumBuffers > 0);
+
+    mpExchangeMgr = apMgr;
+
+    for (j = 0; j < aNumBuffers; j++)
```

Comment:
What does this loop do?

Reply:
From test application, it provides storage, 
https://github.com/project-chip/connectedhomeip/blob/ea07b64987195a3b13663756f148951b17b3e2dc/src/app/tests/integration/chip_im_responder.cpp#L116
for example, gDebugEventBuffer has been saved into void * mpBuffer in LogStorageResources, we need to convert to CircularEventBuffer using the below new CircularEventBuffer and storage saved in mpBuffer(from gDebugEventBuffer)
        https://github.com/project-chip/connectedhomeip/blob/ea07b64987195a3b13663756f148951b17b3e2dc/src/app/reporting/LoggingManagement.cpp#L320

As you see, the overall circular buffer storage is set by application, we hide circular buffer and counter buffer beneath user's view so that user just need provide the storage.

there are 3 buffers for importances, and need to be initialized this way.